daemon/graphdriver/copy/copy.go

```diff
@@ -24,20 +29,61 @@ const (
 	Hardlink
 )
 
-func copyRegular(srcPath, dstPath string, mode os.FileMode) error {
+func copyRegular(srcPath, dstPath string, fileinfo os.FileInfo, copyWithFileRange, copyWithFileClone *bool) error {
 	srcFile, err := os.Open(srcPath)
 	if err != nil {
 		return err
 	}
 	defer srcFile.Close()
 
-	dstFile, err := os.OpenFile(dstPath, os.O_WRONLY|os.O_CREATE, mode)
+	// If the destination file already exists, we shouldn't blow it away
+	dstFile, err := os.OpenFile(dstPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, fileinfo.Mode())
 	if err != nil {
 		return err
 	}
 	defer dstFile.Close()
 
-	_, err = pools.Copy(dstFile, srcFile)
+	if *copyWithFileClone {
+		_, _, err = unix.Syscall(unix.SYS_IOCTL, dstFile.Fd(), ficlone, srcFile.Fd())
+		if err == nil {
+			return nil
+		}
+
+		*copyWithFileClone = false
+		if err == unix.EXDEV {
+			*copyWithFileRange = false
+		}
+	}
+	if *copyWithFileRange {
+		err = doCopyWithFileRange(srcFile, dstFile, fileinfo)
+		// Trying the file_clone may not have caught the exdev case
+		// as the ioctl may not have been available (therefore EINVAL)
```

Comment:
did you mean (therfore ENOSYS)? again, not sure which case the EXDEV is trying to handle. aren't all copies under /var/lib/docker/<driver> which is supposed to be under a single mount?

Reply:
This checks for both exdev and enosys. If FICLONE failed because the ioctl wasn't available, then it falls into this case. I think that checking for EXDEV as well to disable subsequent fast copies is a simple enough check to future-proof us, where it's worth it. 

I can definitely see an admin moving layers "by hand" to another device to spread space usage.

Reply:
Got you. I misread the comment. Nothing wrong with the code. I was just confused because you mentioned the case of kernel with no clone ioctl, but not the common case of fs with no clone support (e.g. ext4) which would return EOPNOTSUPP

Reply:
So, good to go?

Reply:
LGTM